connector_salesforce/unit/exporter_synchronizer.py

```diff
@@ -0,0 +1,354 @@
+from openerp.addons.connector.unit.synchronizer import ExportSynchronizer
+from ..unit.rest_api_adapter import with_retry_on_expiration
+
+_logger = logging.getLogger('salesforce_export_synchronizer')
+
+
+class SalesforceExportSynchronizer(ExportSynchronizer):
+    """Exporter to export an Odoo record to Salesforce.
+
+    Exporting using a Salesforce external id field is supported and
+    should  be use only if model is exported only.
+    It is done by setting _sf_lookup in adapter to the name of external field.
+    Switching from export to import direction is possible when using lookup
+    but not the way around. To be able to do this all external fields
+    must be field with SF Object Id.
+    """
+
+    def __init__(self, connector_env):
+        super(SalesforceExportSynchronizer, self).__init__(connector_env)
+        self.salesforce_id = None
+        self.binding_record = None
+        self.binding_id = None
+
+    def _after_export(self, binding_id):
+        """ Hook called after the export"""
+        return
+
+    def _before_export(self):
+        """ Hook called before the export"""
+        return
+
+    def _deactivate(self):
+        """Implementation of the deactivation
+        of an Odoo record on Salesforce
+        """
+        # external id not supported in delete
+        # In Salesforce nothing is deleted it is just a flag
+        # And deleted ind continue to exist into recycle bin.
+        # Recycle bin lifespan is a Salesforce parameter
+        # by default it is 15 days
+        assert self.binding_record.salesforce_id
+        if self.backend_adapter.exists(self.salesforce_id):
+            self.backend_adapter.delete(self.binding_record.salesforce_id)
+
+    def _to_deactivate(self):
+        """Predicate that decide if an Odoo record
+        must be deactivated on Salesforce
+        """
+        assert self.binding_record
+        model = self.session.pool[self.model._name]
+        cols = set(model._columns.keys())
+        cols.update(model._inherit_fields.keys())
+        if 'active' in cols:
+            if not self.binding_record.active:
+                return True
+        return False
+
+    def _get_record(self):
+        """Return current Odoo record
+
+        :return: an record of Odoo model base on
+                 `_model_name` key
+        :rtype: :py:class:`openerp.osv.orm.BrowseRecord`
+        """
+        assert self.binding_id
+        return self.session.browse(
+            self._model_name,
+            self.binding_id
+        )
+
+    def _map_data_for_upsert(self, mapper, **kwargs):
+        """ Call the convert function of the Mapper
+
+        in order to get converted record by using
+        mapper.data or mapper.data_for_create
+
+        :return: mapped dic of data to be used by
+                 :py:meth:``models.Model.create``
+        :rtype: dict
+        """
+        data = mapper.values(**kwargs)
+        self._validate_data(data)
```

Comment:
Calling `self._validate_data(data)` here could be a problem if you want to override `_map_data_for_upsert` at some place if a mapper returns incomplete data that you complete with values after `super` in you override. The validation will be between the initial mapper and your override.